src/lib/support/PersistentArray.h

```diff
@@ -0,0 +1,316 @@
+            VerifyOrReturnError(this->mCount < this->mLimit, CHIP_ERROR_INVALID_ARGUMENT);
+            ReturnErrorOnFailure(this->Copy(mEntries[this->mCount++], value));
+            ReturnErrorOnFailure(this->Save());
+        }
+        OnEntryAdded(value);
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * Stores a new entry. The Compare() method is used to ensure that the entry is unique.
+     * If the new entry matches an existing entry, do_override must be true, otherwise
+     * CHIP_ERROR_DUPLICATE_KEY_ID is returned.
+     * @param index Zero-based position within the table.
+     * @param entry Value to be store in the table.
+     * @return CHIP_NO_ERROR on success. CHIP_ERROR_DUPLICATE_KEY_ID is the entry is
+     * already stored in the table. CHIP_ERROR_INVALID_ARGUMENT if index >= limit.
+     */
+    CHIP_ERROR Set(const EntryType & value, bool do_override = true)
+    {
+        CHIP_ERROR err = this->Load();
+        VerifyOrReturnError(CHIP_ERROR_NOT_FOUND == err || CHIP_NO_ERROR == err, err);
+
+        // Check existing
+        for (size_t i = 0; i < this->mCount; ++i)
+        {
+            EntryType & e = mEntries[i];
+            if (this->Compare(value, e))
+            {
+                // Already registered
+                VerifyOrReturnError(do_override, CHIP_ERROR_DUPLICATE_KEY_ID);
+                ReturnErrorOnFailure(this->Copy(e, value));
+                return this->Save();
+            }
+        }
+        // Insert last
+        VerifyOrReturnError(this->mCount < this->mLimit, CHIP_ERROR_INVALID_LIST_LENGTH);
+        ReturnErrorOnFailure(this->Copy(mEntries[this->mCount++], value));
+        ReturnErrorOnFailure(this->Save());
+        OnEntryAdded(value);
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Returns the stored entry at a given index.
+     * @param index Zero-based position within the table.
+     * @param entry On success, contains the value matching the given index.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if index is greater than the index of the last entry on the table.
+     */
+    CHIP_ERROR Get(size_t index, EntryType & value)
+    {
+        ReturnErrorOnFailure(this->Load());
+        VerifyOrReturnError(index < this->mLimit, CHIP_ERROR_NOT_FOUND);
+        VerifyOrReturnError(index < this->mCount, CHIP_ERROR_NOT_FOUND);
+        ReturnErrorOnFailure(this->Copy(value, mEntries[index]));
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Returns the value matching the given entry. The Compare() method
+     *        is used to identify the matching entry within the table.
+     * @param index Zero-based position within the table.
+     * @param entry Used as an input for comparison, and output for the rest of the values.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if no entry matches the given input.
+     */
+    CHIP_ERROR Get(EntryType & value, bool do_create = false)
```

Comment:
So here `value` is an inout parameter, right?  It needs to be _very_ clearly documented in terms of what fields need to be set to what on the "in" side and what state it will put things in when do_create is true.

Reply:
I have improved the documentation on this function.